apps/emqx/src/bpapi/emqx_bpapi_trans.erl

```diff
@@ -0,0 +1,195 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @hidden This parse transform generates BPAPI metadata function for
+%% a module, and helps dialyzer typechecking RPC calls
+-module(emqx_bpapi_trans).
+
+-export([parse_transform/2, format_error/1]).
+
+%%-define(debug, true).
+
+-define(META_FUN, bpapi_meta).
+
+-type semantics() :: call | cast.
+
+-record(s,
+        { api              :: atom()
+        , module           :: atom()
+        , version          :: non_neg_integer() | undefined
+        , targets = []     :: [{semantics(), emqx_bpapi:call(), emqx_bpapi:call()}]
+        , errors = []      :: list()
+        , file
+        }).
+
+format_error(invalid_name) ->
+    "BPAPI module name should follow <API>_proto_v<number> pattern";
+format_error({invalid_fun, Name, Arity}) ->
+    io_lib:format("malformed function ~p/~p. "
+                  "BPAPI functions should have exactly one clause "
+                  "and call (emqx_|e)rpc at the top level",
+                  [Name, Arity]).
+
+parse_transform(Forms, _Options) ->
+    log("Original:~n~p", [Forms]),
+    State = #s{file = File} = lists:foldl(fun go/2, #s{}, Forms),
+    log("parse_trans state: ~p", [State]),
+    case check(State) of
+        [] ->
+            finalize(Forms, State);
+        Errors ->
+            {error, [{File, [{Line, ?MODULE, Msg} || {Line, Msg} <- Errors]}], []}
+    end.
+
+%% Scan erl_forms:
+go({attribute, _, file, {File, _}}, S) ->
+    S#s{file = File};
+go({attribute, Line, module, Mod}, S) ->
+    case api_and_version(Mod) of
+        {ok, API, Vsn} -> S#s{api = API, version = Vsn, module = Mod};
+        error          -> push_err(Line, invalid_name, S)
+    end;
+go({function, _Line, introduced_in, 0, _}, S)  ->
```

Comment:
Just curious: this is no longer required and parsed?  I believe it was in the internal state before. :thinking:

Reply:
I changed the implementation from annotations to behaviour callbacks last night. It simplifies the parse transform, which is always good, since metaprogramming is tedious.